File: gettags.pl
#!/usr/bin/env perl

use Image::Magick;
use Music::Tag;

# set these to your Amazon Developer keys
my $token = '';
my $secret_key = '';

$opt_f = shift @ARGV;
$opt_i = shift @ARGV;
exit unless defined $opt_f && defined $opt_i;

my $info = Music::Tag->new($opt_f, {
	'quiet' => 1,
	,'secret_key' => $secret_key
	,'token' => $token
});

$info->add_plugin('Amazon') if $token;
$info->get_tag();

my $picture = undef;
if($info->picture_exists()) {
        $picture = $info->get_picture();
} else {
	# the Music::Tag::MP3 looks for APIC 'Cover (front)', and none of my
	# MP3s have that specific type of cover set.  So, if it's an MP3, grab
	# the first APIC if it exists.
        if($info->filetype() eq 'mp3' && exists($info->plugin('MP3')->mp3()->{'ID3v2'})) {
                $picture = $info->plugin('MP3')->mp3()->{'ID3v2'}->frame_select_by_descr("APIC");
        }
}

# print the information
printf("{title:\"%s\",artist:\"%s\",album:\"%s\",length:%u,cover:%u,url:\"%s\"}\n",
  $info->has_title() ? quote($info->get_title()) : '',
  $info->has_artist() ? quote($info->get_artist()) : '',
  $info->has_album() ? quote($info->get_album()) : '',
  $info->has_duration() ? $info->get_duration() : 0,
  defined($picture) ? 1 : 0,
  $info->has_url() ? $info->get_url() : '',
);

# write out the photo
if($picture) {
  my $image = new Image::Magick;
  $image->BlobToImage($picture->{'_Data'});
  $image->Resize('height'=>256,'width'=>256);
  $image->Write('static/images/'.$opt_i.'.jpg');
}

$info->close();

sub quote {
  my ($string) = $_[0];
  $string =~ s/&/&amp;/g;
  $string =~ s/"/&quot;/g;
  $string =~ s/</&lt;/g;
  $string =~ s/>/&gt;/g;
  return $string;
}

File: gettags.py
#!/usr/bin/env python

from cgi import escape
from math import floor
from mutagen import File
from mutagen.easyid3 import EasyID3
from os import unlink
from pprint import pprint
from PythonMagick import Image, Blob
from tempfile import NamedTemporaryFile
import json
import sys

def extractCover(uuid, data):
	"""Extract the cover, resize, and save."""
	f = NamedTemporaryFile(delete=False)
	f.write(data)
	f.close()
	image = Image(f.name)
	image.magick('JPEG') 
	image.sample('256x256')
	image.write('htdocs/images/'+uuid+'.jpg')
	unlink(f.name)

if len(sys.argv) > 1:
	fname = sys.argv[1]
	uuid = sys.argv[2]
else:
	exit(0)

data = {}
audio = File(fname, easy=True)
if hasattr(audio, 'mime') == False:
	exit(0)

for f in set(['title','artist','album']):
	if audio.has_key(f): data[f] = escape(audio[f][0])
data['length'] = int(floor(audio.info.length*1000))

# now try to extract the cover.
audio = File(fname, easy=False)

# each filetype has a different method
if 'audio/vorbis' in audio.mime or 'audio/x-flac' in audio.mime:
	for f in set(['title','artist','album']):
		if audio.has_key(f): data[f] = escape(audio[f][0])
	if audio.has_key('pictures'):
		data['cover'] = uuid+'.jpg';
		extractCover(uuid, audio['pictures'][0])

if 'audio/aac' in audio.mime:
	if audio.has_key('covr'):
		data['cover'] = uuid+'.jpg';
		extractCover(uuid, audio['covr'][0])

if 'audio/mp3' in audio.mime:
	if audio.has_key('TIT2'): data['title'] = escape(audio['TIT2'][0])
	if audio.has_key('TPE1'): data['artist'] = escape(audio['TPE1'][0])
	if audio.has_key('TALB'): data['album'] = escape(audio['TALB'][0])
	if audio.has_key('APIC:'):
		data['cover'] = uuid+'.jpg';
		extractCover(uuid, audio['APIC:'].data)

print json.dumps(data)
